File: ios/BpWidget/BpSharedStore.swift
import Foundation

enum BpSharedStore {
    static let appGroupId = "group.com.bangguoxiong.bpCompanion"
    static let recordsKey = "bp_records_v1"
    static let lastMessageKey = "widget_last_message"

    private static var defaults: UserDefaults? {
        UserDefaults(suiteName: appGroupId)
    }

    struct StoredRecord: Codable {
        let id: String
        let systolic: Int
        let diastolic: Int
        let pulse: Int?
        let time: String
        let context: Int
        let note: String
    }

    static func loadRecords() -> [StoredRecord] {
        guard let raw = defaults?.string(forKey: recordsKey),
              let data = raw.data(using: .utf8),
              let records = try? JSONDecoder().decode([StoredRecord].self, from: data) else {
            return []
        }
        return records
    }

    static func latestRecord() -> StoredRecord? {
        loadRecords().sorted { $0.time > $1.time }.first
    }

    static func appendRecord(systolic: Int, diastolic: Int, pulse: Int?) throws {
        guard systolic >= 50, systolic <= 300, diastolic >= 30, diastolic <= 200 else {
            throw NSError(domain: "BpSharedStore", code: 1)
        }

        var records = loadRecords()
        let now = Date()
        let context: Int
        let hour = Calendar.current.component(.hour, from: now)
        if hour < 12 {
            context = 0
        } else if hour >= 21 {
            context = 1
        } else {
            context = 4
        }

        let id = "\(Int(now.timeIntervalSince1970 * 1_000_000))_\(Int.random(in: 0..<99999))"
        let formatter = ISO8601DateFormatter()

        let record = StoredRecord(
            id: id,
            systolic: systolic,
            diastolic: diastolic,
            pulse: pulse,
            time: formatter.string(from: now),
            context: context,
            note: ""
        )
        records.append(record)

        let data = try JSONEncoder().encode(records)
        guard let json = String(data: data, encoding: .utf8) else { return }
        defaults?.set(json, forKey: recordsKey)
        defaults?.set("已保存 \(systolic)/\(diastolic)", forKey: lastMessageKey)
    }
}

File: android/app/src/main/kotlin/com/bpcompanion/bp_companion/BpSharedStore.kt
package com.bpcompanion.bp_companion

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant
import java.time.format.DateTimeFormatter
import kotlin.random.Random

object BpSharedStore {
    const val PREFS_NAME = "HomeWidgetPreferences"
    const val RECORDS_KEY = "bp_records_v1"
    const val DRAFT_SYS_KEY = "widget_draft_sys"
    const val DRAFT_DIA_KEY = "widget_draft_dia"
    const val DRAFT_PULSE_KEY = "widget_draft_pulse"
    const val LAST_MESSAGE_KEY = "widget_last_message"

    fun loadRecords(context: Context): JSONArray {
        val raw = prefs(context).getString(RECORDS_KEY, null) ?: return JSONArray()
        return try {
            JSONArray(raw)
        } catch (_: Exception) {
            JSONArray()
        }
    }

    fun latestRecord(context: Context): JSONObject? {
        val records = loadRecords(context)
        if (records.length() == 0) return null
        var latest: JSONObject? = null
        for (i in 0 until records.length()) {
            val item = records.getJSONObject(i)
            if (latest == null || item.getString("time") > latest.getString("time")) {
                latest = item
            }
        }
        return latest
    }

    fun appendRecord(context: Context, systolic: Int, diastolic: Int, pulse: Int?) {
        require(systolic in 50..300 && diastolic in 30..200)
        val records = loadRecords(context)
        val now = Instant.now()
        val hour = now.atZone(java.time.ZoneId.systemDefault()).hour
        val measureContext = when {
            hour < 12 -> 0
            hour >= 21 -> 1
            else -> 4
        }
        val record = JSONObject()
            .put("id", "${System.currentTimeMillis()}_${Random.nextInt(0, 99999)}")
            .put("systolic", systolic)
            .put("diastolic", diastolic)
            .put("pulse", pulse)
            .put("time", DateTimeFormatter.ISO_INSTANT.format(now))
            .put("context", measureContext)
            .put("note", "")
        records.put(record)
        prefs(context).edit()
            .putString(RECORDS_KEY, records.toString())
            .putString(LAST_MESSAGE_KEY, "已保存 $systolic/$diastolic")
            .apply()
    }

    fun getDraft(context: Context, key: String, default: Int): Int =
        prefs(context).getInt(key, default)

    fun setDraft(context: Context, key: String, value: Int) {
        prefs(context).edit().putInt(key, value).apply()
    }

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}
